File: app/services/ticket_service.py
from datetime import datetime, timedelta
from typing import List

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from .. import models, schemas
from ..locks import ticket_lock
from ..ws_manager import manager

HOLD_DURATION_MINUTES = 10


async def hold_ticket(db: Session, data: schemas.TicketHoldRequest, user: models.User) -> schemas.TicketHoldResponse:
    with ticket_lock(data.ticket_id):
        ticket = db.query(models.Ticket).filter(models.Ticket.id == data.ticket_id).first()
        if ticket is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Ingresso não encontrado")
        if ticket.status != models.TicketStatus.AVAILABLE:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Ingresso indisponível. Status atual: {ticket.status.value}",
            )

        ticket.status = models.TicketStatus.RESERVED
        ticket.reserved_at = datetime.utcnow()
        ticket.reserved_by = user.id
        ticket.version = (ticket.version or 0) + 1
        db.commit()
        db.refresh(ticket)

    await _broadcast_availability(db, ticket)

    return schemas.TicketHoldResponse(
        ticket_id=ticket.id,
        seat_number=ticket.seat_number,
        price=ticket.price,
        hold_expiration=datetime.utcnow() + timedelta(minutes=HOLD_DURATION_MINUTES),
    )


async def release_ticket_hold(db: Session, ticket_id: str) -> None:
    with ticket_lock(ticket_id):
        ticket = db.query(models.Ticket).filter(models.Ticket.id == ticket_id).first()
        if ticket is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Ingresso não encontrado")

        if ticket.status != models.TicketStatus.RESERVED:
            return

        ticket.status = models.TicketStatus.AVAILABLE
        ticket.reserved_at = None
        ticket.reserved_by = None
        ticket.version = (ticket.version or 0) + 1
        db.commit()
        db.refresh(ticket)

    await _broadcast_availability(db, ticket)


async def confirm_ticket_sale(db: Session, ticket_id: str, order_id: str) -> None:
    with ticket_lock(ticket_id):
        ticket = db.query(models.Ticket).filter(models.Ticket.id == ticket_id).first()
        if ticket is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Ingresso não encontrado")
        if ticket.status != models.TicketStatus.RESERVED:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Ingresso precisa estar reservado antes da venda")

        ticket.status = models.TicketStatus.SOLD
        ticket.order_id = order_id
        db.commit()
        db.refresh(ticket)

    await _broadcast_availability(db, ticket)


def get_available_tickets(db: Session, event_id: str) -> List[schemas.TicketResponse]:
    tickets = (
        db.query(models.Ticket)
        .filter(models.Ticket.event_id == event_id, models.Ticket.status == models.TicketStatus.AVAILABLE)
        .order_by(models.Ticket.seat_number)
        .all()
    )
    return [
        schemas.TicketResponse(id=t.id, seat_number=t.seat_number, price=t.price, status=t.status.value)
        for t in tickets
    ]


def find_expired_reservations(db: Session) -> List[models.Ticket]:
    expiration_time = datetime.utcnow() - timedelta(minutes=HOLD_DURATION_MINUTES)
    return (
        db.query(models.Ticket)
        .filter(models.Ticket.status == models.TicketStatus.RESERVED, models.Ticket.reserved_at < expiration_time)
        .all()
    )


async def _broadcast_availability(db: Session, ticket: models.Ticket) -> None:
    available_count = (
        db.query(models.Ticket)
        .filter(models.Ticket.event_id == ticket.event_id, models.Ticket.status == models.TicketStatus.AVAILABLE)
        .count()
    )
    update = schemas.TicketAvailabilityUpdate(
        event_id=ticket.event_id,
        ticket_id=ticket.id,
        seat_number=ticket.seat_number,
        status=ticket.status.value,
        available_count=available_count,
    )
    await manager.broadcast(ticket.event_id, update.model_dump())

File: src/services/ticketService.js
'use strict';
const TicketModel = require('../models/Ticket');
const { TicketStatus } = require('../models/enums');
const { HttpError } = require('../utils/errors');
const { withTicketLock } = require('../utils/locks');
const { manager } = require('../utils/wsManager');

const HOLD_DURATION_MINUTES = 10;

async function holdTicket(data, user) {
  const ticket = await withTicketLock(data.ticket_id, () => {
    const t = TicketModel.findById(data.ticket_id);
    if (!t) {
      throw new HttpError(404, 'Ingresso não encontrado');
    }
    if (t.status !== TicketStatus.AVAILABLE) {
      throw new HttpError(409, `Ingresso indisponível. Status atual: ${t.status}`);
    }

    return TicketModel.updateStatus(t.id, {
      status: TicketStatus.RESERVED,
      reservedAt: new Date().toISOString(),
      reservedBy: user.id,
    });
  });

  await broadcastAvailability(ticket);

  return {
    ticket_id: ticket.id,
    seat_number: ticket.seat_number,
    price: Number(ticket.price),
    hold_expiration: new Date(Date.now() + HOLD_DURATION_MINUTES * 60 * 1000).toISOString(),
  };
}

async function releaseTicketHold(ticketId) {
  const ticket = await withTicketLock(ticketId, () => {
    const t = TicketModel.findById(ticketId);
    if (!t) {
      throw new HttpError(404, 'Ingresso não encontrado');
    }
    if (t.status !== TicketStatus.RESERVED) {
      return null; // nada a fazer, igual ao early-return do Python
    }

    return TicketModel.updateStatus(t.id, {
      status: TicketStatus.AVAILABLE,
      reservedAt: null,
      reservedBy: null,
    });
  });

  if (ticket) {
    await broadcastAvailability(ticket);
  }
}

async function confirmTicketSale(ticketId, orderId) {
  const ticket = await withTicketLock(ticketId, () => {
    const t = TicketModel.findById(ticketId);
    if (!t) {
      throw new HttpError(404, 'Ingresso não encontrado');
    }
    if (t.status !== TicketStatus.RESERVED) {
      throw new HttpError(409, 'Ingresso precisa estar reservado antes da venda');
    }

    return TicketModel.updateStatus(t.id, {
      status: TicketStatus.SOLD,
      orderId,
      bumpVersion: false,
    });
  });

  await broadcastAvailability(ticket);
}

function getAvailableTickets(eventId) {
  return TicketModel.findAvailableByEvent(eventId).map((t) => ({
    id: t.id,
    seat_number: t.seat_number,
    price: Number(t.price),
    status: t.status,
  }));
}

function findExpiredReservations() {
  const expirationTime = new Date(
    Date.now() - HOLD_DURATION_MINUTES * 60 * 1000
  ).toISOString();
  return TicketModel.findExpiredReservations(expirationTime);
}

async function broadcastAvailability(ticket) {
  const availableCount = TicketModel.countAvailableByEvent(ticket.event_id);
  const update = {
    event_id: ticket.event_id,
    ticket_id: ticket.id,
    seat_number: ticket.seat_number,
    status: ticket.status,
    available_count: availableCount,
  };
  manager.broadcast(ticket.event_id, update);
}

module.exports = {
  HOLD_DURATION_MINUTES,
  holdTicket,
  releaseTicketHold,
  confirmTicketSale,
  getAvailableTickets,
  findExpiredReservations,
};
